Upgrade to C++17 std::filesystem library

jmarrec opened the pull request:
Pull request overview
---------------------



 - Fixes #8376
 - Original NFP was in https://github.com/NREL/EnergyPlus/pull/8411, but NFP doc was merged into this PR


### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

jmarrec committed:
Make everything (almost) use std::filesystem, but do not use (yet) fs::path everywhere instead of std::string (and +)

jmarrec committed:
NFP for C++17 std::filesystem

jmarrec committed:
Expand unit tests

jmarrec committed:
Do changes and leave TODO for potential removals

jmarrec committed:
Renable previously failing tests

jmarrec committed:
Save WIP: it compiles. Unit tests are ok. But integration are all failing, so probably an error in CommandLineInterface or somewhere early like that

jmarrec committed:
Found the issue: fs::path::replace_extension mutates. Not sure if I should provide a helper or just ensure I make copies before

jmarrec committed:
Caught a few I missed

jmarrec committed:
I made changes to src/ConvertInputFormat without realizaing... MSVC throws because this project uses CXX11 and not C++17

jmarrec committed:
Thanks MSVC: cannot convert argument 1 from 'std::filesystem::path' to 'const std::string &'

I think this is going to be fun

jmarrec committed:
Port sqlite to use fs::path for dbName since that is what's passed, and msvc complains

jmarrec committed:
Use fs::path in the tablesFiles for BtwxtManager

jmarrec committed:
Port PluginManager to use fs::path

jmarrec committed:
OutputTabular to fs::path

jmarrec committed:
try to make MSVC happy given that it doesn't go as easily from/to fs::path/std::string

jmarrec committed:
ok msvc compiles

jmarrec committed:
Make PythonCopyStandardLib a bit smarter. I got bit REAL BAD by the fact that it doesn't check if you switched python version

Just Spent 5 or 6 hours trying to debug why I couldn't import python plugin thinking it was fs::path related, when really I have built wiht python3.8 at one point and later changed to python3.9 (and I did that on TWO machines, so I didn't catch the problem), so the script copied the lib-dynload 3.8 so files, and never redid it later. So my 3.9 couldn't find _ctypes because there was no file with the right ABI

```
   ** Severe  ** Failed to import module "PythonPluginCustomOutputVariable"
   **   ~~~   ** Python error description follows: 
   **   ~~~   ** ModuleNotFoundError("No module named '_ctypes'")
   **   ~~~   ** Python traceback follows: 
   **   ~~~   ** ```
   **   ~~~   **  >>> Traceback (most recent call last):
   **   ~~~   **  >>>   File "../EnergyPlus/testfiles/PythonPluginCustomOutputVariable.py", line 2, in <module>
    from pyenergyplus.plugin import EnergyPlusPlugin
   **   ~~~   **  >>>   File "/home/julien/Software/Others/EnergyPlus-build/Products/pyenergyplus/plugin.py", line 3, in <module>
    from pyenergyplus.api import EnergyPlusAPI
   **   ~~~   **  >>>   File "/home/julien/Software/Others/EnergyPlus-build/Products/pyenergyplus/api.py", line 1, in <module>
    from ctypes import cdll, c_char_p, c_void_p
   **   ~~~   **  >>>   File "/home/julien/Software/Others/EnergyPlus-build/Products/python_standard_lib/ctypes/__init__.py", line 7, in 
```

jmarrec committed:
Use fs::path in ExternalInterface internally.

I'm questioning whether it makes sense

jmarrec committed:
Mod unit tests after rebase.

jmarrec committed:
Merge remote-tracking branch 'origin/nfp_filesystem' into 8376_FileSystem_Move

jmarrec commented on cmake/PythonCopyStandardLib.py:
```diff
@@ -25,12 +26,37 @@
 ctypes_package_dir = os.path.dirname(ctypes_import_file)
 standard_lib_dir = os.path.dirname(ctypes_package_dir)
+
+if os.path.exists(target_dir):
+    # Let's check the library files to see if the ABI matches
+    # Otherwise if you build with say python 3.8 initially, and then switch to
+    # python 3.9, your lib-dynload will still have the 38 .so files
+    std_lib_ctypes_sos = gb.glob(os.path.join(standard_lib_dir, "**/_ctypes.*"), recursive=True)
+    this_lib_ctypes_sos = gb.glob(os.path.join(target_dir, "**/_ctypes.*"), recursive=True)
+
+    def find_libs(dirPath):
+        sos = []
+        for ext in ['a', 'so', 'lib']:
+            sos += gb.glob(os.path.join(dirPath, "**/*.{}*".format(ext)),
+                           recursive=True)
+        return [os.path.basename(f) for f in sos]
+
+    std_lib_ctypes_sos = find_libs(standard_lib_dir)
+    this_lib_ctypes_sos = find_libs(target_dir)
+
+    if ((set(std_lib_ctypes_sos) - set(this_lib_ctypes_sos)) or
+        (set(this_lib_ctypes_sos) - set(std_lib_ctypes_sos))):
+        print("Detected changes in the python libs, wiping and recopying")
+        shutil.rmtree(target_dir)
+    else:
+        # File names match
+        sys.exit(0)
```
I got burned pretty badly by this script. I built with python 3.8 originally. Then I changed in the cmake to python3.9, but the shared library files didn't get copied again, so when running PythonPlugin example files, my python3.9 couldn't load the standard library in `python_standard_library/lib-dynload` since they were the 3.8 ones... 

5 hours later I realized this was the problem and not my fs::path changes.

Myoldmopar replied:
Sorry, yeah, that definitely needed some love.  I haven't looked at your changes yet, but if you need me to make some updates there still I'm happy to do it.

Myoldmopar replied:
I see in the code now that you added this block to do a very intelligent check on whether to update the library.  This looks great!

Myoldmopar replied:
Still looks great :smile:

jmarrec committed:
Adjust linker calls for Clang

jmarrec commented on src/ConvertInputFormat/CMakeLists.txt:
```diff
@@ -1,7 +1,7 @@
 project(ConvertInputFormat)
 
-set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD 17)
```
I started touching this subfolder without realizing (...), so instead of reverting the changes I just bumped to C++17.

jmarrec commented on src/EnergyPlus/CMakeLists.txt:
```diff
@@ -767,37 +767,61 @@ endif()
+
+# See https://en.cppreference.com/w/cpp/filesystem#Notes
+if ((CMAKE_COMPILER_IS_GNUCXX AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 9.1) OR
+    ( ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang" OR "${CMAKE_CXX_COMPILER_ID}" STREQUAL "AppleClang") AND (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 9.0)))
```
Some shenanigans that are needed for GCC < 9.1 or Clang < 9 to link to std::filesystem.

jmarrec commented on src/EnergyPlus/CommandLineInterface.cc:
```diff
@@ -238,40 +255,41 @@ namespace CommandLineInterface {
+        DataStringGlobals::inputFilePathNameOnly = FileSystem::removeFileExtension(FileSystem::getFileName(DataStringGlobals::inputFilePath));
+        DataStringGlobals::inputDirPath = FileSystem::getParentDirectoryPath(DataStringGlobals::inputFilePath);
+
+        // TODO: would this be better?
+        // FileSystem::InputFileType fileType = FileSystem::getInputFileType(DataStringGlobals::inputFilePath);
```
TODO: not sure if worth it to provide a storngly type enumeration or not.

jmarrec replied:
Note: similar block of code can be found in ConvertInputFormat/main.cpp

Myoldmopar replied:
I mean, that would certainly be great I think.  I can't think of a reason why we would need these individual flags inside E+.  @mbadams5 ?

jmarrec commented on src/EnergyPlus/CommandLineInterface.cc:
```diff
@@ -322,143 +331,168 @@ namespace CommandLineInterface {
+        state.files.inStatFilePath.filePath = state.files.inputWeatherFilePath.filePath;
+        state.files.inStatFilePath.filePath.replace_extension(".stat");
+        // Or is it better to provide a helper like this?
+        // state.files.inStatFilePath.filePath = FileSystem::replaceFileExtension(state.inputWeatherFilePath.filePath, ".stat");
```
TODO: couldn't really decide whether the helper is a good idea, or leave the burden of not messing it up to the developper?

Myoldmopar replied:
Are you saying you could accidentally stomp on a previous value because it is mutating the underlying reference?  I don't think a helper should be really necessary, primarily because if a developer messes with that the wrong way, the simulation and the CI system should complain with huge red badges alerting that things went terribly wrong.  But, all that said, if a worker function can help avoid that ever happening, then OK, maybe it's worth it.

jmarrec replied:
It mutates the path it's called on yes: https://en.cppreference.com/w/cpp/filesystem/path/replace_extension . Signature is `path& replace_extension( const path& replacement = path() );`

See an example on https://godbolt.org/z/Mvze3W

```c++
#include <filesystem>
#include <iostream>

namespace fs = std::filesystem;

int main() {

    fs::path inputWeatherFilePath("/path/to/myfile.epw");
    fs::path inStatFilePath = inputWeatherFilePath.replace_extension(".stat");
    std::cout << "inputWeatherFilePath=" << inputWeatherFilePath.string() << '\n';
    // !! => inputWeatherFilePath=/path/to/myfile.stat
    std::cout << "inStatFilePath=" << inStatFilePath.string() << '\n';

}
```

The thing with having a helper function is that nothing prevents anyone from **not** using it and directly calling the `fs::path::replace_extension` instead of the helper...

jmarrec commented on src/EnergyPlus/CommandLineInterface.cc:
```diff
@@ -322,143 +331,168 @@ namespace CommandLineInterface {
-        state.files.inStatFileName.fileName = weatherFilePathWithoutExtension + ".stat";
-        eplusADSFileName = inputDirPathName + "eplusADS.inp";
+
+        state.files.screenCsv.filePath = composePath(screenSuffix + ".csv");
+        // TODO, why are these relative paths?
+        state.files.delightIn.filePath = "eplusout.delightin";
+        DataStringGlobals::outputDelightOutFilePath = "eplusout.delightout";
+
+        // TODO: why is this relative?
+        state.files.iniFile.filePath = "Energy+.ini";
```
I'm not sure I understand why these were different? I guess for iniFile I do, it's because it'll look for the Energy+.ini in the current directory or next to the IDF later, but the delightout?

Myoldmopar replied:
:shrug:

jmarrec commented on src/EnergyPlus/ConfiguredFunctions.hh:
```diff
@@ -49,12 +49,13 @@
 namespace EnergyPlus {
 
-std::string configured_source_directory();
+fs::path configured_source_directory();
```
TODO: shouldn't these two be made `inline` (and move to header-only) (or at least `static`)?

Myoldmopar replied:
Yes, they should.  That was added in a bit of a rush, and ended up the way they are because they needed to be fixed in a hurry.  If you want to clean it up that is totally fine.

Myoldmopar replied:
They can just be `constexpr char *` variables now actually, which is even simpler.

jmarrec replied:
If I make them `constexpr char *` then all calls would have to look like `fs:path(configured_source_directory()) / xxx`

lefticus replied:
`constexpr string_view` would be better than `constexpr char *` but in either case, yes you'd have to call your conversion function to use them.

jmarrec committed:
Build error on windows (establishclientsocket doesn't accept fs::string_type) and minor adjustments

jmarrec committed:
Merge remote-tracking branch 'refs/remotes/origin/develop'

Conflicts:
	src/EnergyPlus/CMakeLists.txt
	tst/EnergyPlus/unit/UnitaryHybridAirConditioner.unit.cc

jmarrec committed:
Copy paste errors introduced in a506dced46397541da76372cd85db60b19e92cc3

jmarrec committed:
Adjust new unit test to conform to std::filesystem changes

Myoldmopar commented:
Obviously CI isn't quite happy with you yet on this one, but I'm excited to see the progress.  Let me know if I can help with any of the issues you've identified.

Myoldmopar commented on scripts/dev/find_byref_bool_override.py:
```diff
@@ -111,13 +111,6 @@
         ]
     },
 
-    # Could be changed to just return a string
```
One less to check!

Myoldmopar commented on src/EnergyPlus/CMakeLists.txt:
```diff
@@ -765,37 +765,63 @@ endif()
+         Windows-CalcEngine
+         airflownetworklib
+         ssc
+)
```
I do like cleanups in the cmake.  I'd like to know how badly this will conflict with the changes in the work from @lefticus  to modernize our cmake (#8425).  I'm not saying you have to change your approach here, I'm just wanting to prepare myself mentally :laughing:

jmarrec replied:
I have seen #8425 and aligned my use of PUBLIC/PRIVATE attributes to match, so it shouldn't be that bad. (The format of this block (switching to one per line) is also copied over from lefticus's PR)

Plus, I'm touching like 4 or 5 blocks in the CMakeLists.txts, so the merge conflicts should be fine. I'd be happy to fix conflicts on whichever PR is merged second.

Myoldmopar commented on src/EnergyPlus/CommandLineInterface.cc:
```diff
@@ -169,15 +167,32 @@ namespace CommandLineInterface {
 
-        if (!opt.isSet("-i") && !legacyMode) inputIddFileName = exeDirectory + inputIddFileName;
+        if (!opt.isSet("-i") && !legacyMode) {
+            DataStringGlobals::inputIddFilePath = DataStringGlobals::exeDirectoryPath / DataStringGlobals::inputIddFilePath;
```
:heart_eyes: I know it's silly, but this is just nice.  It's the same as when I started using `pathlib` in Python.  This just makes so much sense.

Myoldmopar commented on src/EnergyPlus/CommandLineInterface.cc:
```diff
@@ -322,143 +331,168 @@ namespace CommandLineInterface {
+        // Helper to construct output file path
+        auto composePath = [&outputFilePrefixFullPath](const std::string& suffix) -> fs::path {
+            return fs::path(outputFilePrefixFullPath.string() + suffix);
+        };
```
:+1:

Myoldmopar commented on src/EnergyPlus/CommandLineInterface.cc:
```diff
@@ -501,36 +535,39 @@ namespace CommandLineInterface {
+                //DataStringGlobals::CurrentWorkingFolder = "";
+            //} else {
+                //DataStringGlobals::CurrentWorkingFolder.erase(TempIndx + 1);
+            //}
```
:shrug:

Myoldmopar commented on src/EnergyPlus/ExternalInterface.cc:
```diff
@@ -1120,12 +1123,16 @@ namespace ExternalInterface {
             // post process as needed in case these are used later
-            FMURootWorkingFolder = "tmp-fmus";
-            FMURootWorkingFolder += pathChar; // getStringFromCharArray( FMUWorkingFolderCharArr );
+            FMURootWorkingFolder = fs::path("tmp-fmus"); // getStringFromCharArray( FMUWorkingFolderCharArr );
```
All these corners of the code...

Myoldmopar commented on src/EnergyPlus/FileSystem.cc:
```diff
@@ -176,93 +188,102 @@ namespace FileSystem {
-        } else
-            return false;
-#endif
+        return fs::exists(filePath) && !fs::is_directory(filePath);
```
This.  :+1: :+1: :+1:

Myoldmopar reviewed (commented):
This is **a lot** of changes.  Really good stuff.  I'm hesitant to approve it because CI is all red, but I don't see any problems from a review of it.  I love where this is going.

jmarrec committed:
Merge branch 'develop' into 8376_FileSystem_Move

Manual conflicts resolved: TARCOGOutput

jmarrec committed:
Minor rename of variable for consistency

jmarrec committed:
Merge error: missed one std::string > fs::path conversion

jmarrec committed:
typo in new test

lefticus commented on src/EnergyPlus/FileSystem.hh:
```diff
@@ -49,43 +49,108 @@
 #include <algorithm>
+#include <string>
+#include <fmt/format.h>
+#if __has_include(<filesystem>)
```
`__has_include` might not necessarily work 100% how we want it to - many compiler vendors provide the headers for features they don't yet support, then give you a compile time warning if you try to use features from that header.

A better option is compiler feature test macros https://en.cppreference.com/w/User:D41D8CD98F/feature_testing_macros#C.2B.2B17 but those are not officially supported until C++20, and MSVC doesn't enable them unless you use C++20 mode (other compilers support them before C++20).

TL;DR, don't change anything, this is probably your best option for today, but I thought it was worth mentioning in case we do have an issue in the future with this.

kbenne replied:
I think cmake's check_include_file might be the best option https://cmake.org/cmake/help/latest/module/CheckIncludeFile.html

jmarrec committed:
Merge remote-tracking branch 'origin/develop' into 8376_FileSystem_Move

Resolve conflicts

jmarrec committed:
Fix merge issue

jmarrec commented:
I merged develop and resolved conflicts. This new warning pops up:

```c++
[ 38%][479/1240] Building CXX object src/EnergyPlus/AirflowNetwork/CMakeFiles/airflownetworklib.dir/src/Properties.cpp.o
In file included from /home/julien/Software/Others/EnergyPlus/src/EnergyPlus/IOFiles.hh:58:0,
                 from /home/julien/Software/Others/EnergyPlus/src/EnergyPlus/DataGlobals.hh:57,
                 from /home/julien/Software/Others/EnergyPlus/src/EnergyPlus/AirflowNetwork/include/AirflowNetwork/../../../Psychrometrics.hh:62,
                 from /home/julien/Software/Others/EnergyPlus/src/EnergyPlus/AirflowNetwork/include/AirflowNetwork/Properties.hpp:52,
                 from /home/julien/Software/Others/EnergyPlus/src/EnergyPlus/AirflowNetwork/src/Properties.cpp:48:
/home/julien/Software/Others/EnergyPlus/src/EnergyPlus/FileSystem.hh:104:64: warning: ‘no_discard’ attribute directive ignored [-Wattributes]
     [[no_discard]] fs::path makeNativePath(fs::path const &path);
                                                                ^
/home/julien/Software/Others/EnergyPlus/src/EnergyPlus/FileSystem.hh:106:65: warning: ‘no_discard’ attribute directive ignored [-Wattributes]
     [[no_discard]] fs::path getFileName(fs::path const &filePath);
```


@lefticus is this equivalent to `set(CXX_STANDARD 17)`? 

https://github.com/NREL/EnergyPlus/blob/e8bf5eab3283afa744b5fa8a5cc6a6a9d5cff10c/CMakeLists.txt#L25

jmarrec commented:
I get the no_discard warnings with gcc 7 and gcc 10.


FYI, There are other unrelated warnings too in Eigen

https://github.com/NREL/EnergyPlus/runs/1888058655#step:6:229

lefticus commented:
> 
> 
> I get the no_discard warnings with gcc 7 and gcc 10.
> 
> FYI, There are other unrelated warnings too in Eigen
> 
> https://github.com/NREL/EnergyPlus/runs/1888058655#step:6:229

It should be `[[nodiscard]]` no `_`

jmarrec commented:
> It should be `[[nodiscard]]` no `_`

:facepalm: . Thank you..

jmarrec committed:
Merge branch 'develop' into 8376_FileSystem_Move

jmarrec committed:
I mistyped nodiscard...

jmarrec committed:
And now that the nodiscard is actually working, it raises this warning

jmarrec committed:
Merge branch 'develop' into 8376_FileSystem_Move

jmarrec committed:
Merge remote-tracking branch 'origin/FixSSCPackaging' into 8376_FileSystem_Move

jmarrec committed:
Merge remote-tracking branch 'origin/develop' into 8376_FileSystem_Move

jmarrec committed:
Merge remote-tracking branch 'origin/develop' into 8376_FileSystem_Move

jmarrec committed:
Fix build error after merge: new unit tests needs to be upgraded to PR

jmarrec committed:
Add a test highlighting the problem that happens on mac currently

jmarrec committed:
Fix getAbsolutePath to resolve mac issue.

On mac, `getProgramPath` :
```
_NSGetExecutablePath, /Users/dev/git/EnergyPlus-build-release/Products/energyplus, pathSize=1024
```

This is a symlink, but points to a relative dir
```
ls -la Products/energyplus
lrwxr-xr-x 1 dev staff 16 Mar  1 17:55 Products/energyplus -> energyplus-9.5.0
```

So witht he previous implementation, fs::absolute is called after, and that resolves to current_dir / energyplus-9.5.0. So unless you called the call from within the exe directory, it failed. THis fixes it

jmarrec committed:
Merge remote-tracking branch 'origin/develop' into 8376_FileSystem_Move

Myoldmopar commented:
Great work @jmarrec !  I will test this out if you feel like you are done with it.

jmarrec commented:
@Myoldmopar I think it's ready for review. the Clang build is failing because it's not targeting 10.15 and getting the `'path' is unavailable: introduced in macOS 10.15` stuff

Myoldmopar commented:
This will be merged right after 9.5 is released to avoid breaking support for 10.14 so close to the release.

nrel-bot-3 commented:
@jmarrec @lgentile it has been 28 days since this pull request was last updated.

jmarrec committed:
Merge develop. This was the worst merge of my life... Expect that I broke a lot of things resolving the several hundred of conflicts

jmarrec committed:
Resolve obvious merge issues. It compiles, tests passed.

jmarrec committed:
More build errors

jmarrec committed:
Unrelated typo

jmarrec committed:
Last build error

jmarrec committed:
Fix paths in tests. Unit tests passing. Integration failing

jmarrec committed:
Another bad merge that explains the failing integration tests (+ typo in test)

jmarrec committed:
Sneaky copy paste error during merge. That should fix the last integration tests

jmarrec committed:
Minor adjustment.

jmarrec committed:
Fix windows build error

jmarrec committed:
Merge remote-tracking branch 'origin/develop' into 8376_FileSystem_Move_update

jmarrec committed:
clang-format-10 (command line used below)

```bash
CHECK_PATH=tst/EnergyPlus/unit
find "$CHECK_PATH" | /bin/grep -E '\.((c|C)c?(pp|xx|\+\+)*$|(h|H)h?(pp|xx|\+\+)*$)' | xargs clang-format-10 -style=file -i -fallback-style=none
CHECK_PATH=src/EnergyPlus
find "$CHECK_PATH" | /bin/grep -E '\.((c|C)c?(pp|xx|\+\+)*$|(h|H)h?(pp|xx|\+\+)*$)' | xargs clang-format-10 -style=file -i -fallback-style=none
```

jmarrec committed:
clang format ConvertInputFormat too

jmarrec committed:
More errors on windows

jmarrec committed:
More errors on windows following merge

jmarrec committed:
Bump aqt qtifw to 4.1 (seems like 4.0 was removed from Qt archive...)

jmarrec committed:
Update test: programPath has .exe on windows

jmarrec committed:
Use unique name for sandbox directory to avoid race conditions when running tests in parallel

jmarrec committed:
Merge remote-tracking branch 'origin/develop' into 8376_FileSystem_Move

jmarrec committed:
Bump decent CI macOS deployment target to 10.15

jmarrec commented:
Test GHA build on my fork at https://github.com/jmarrec/EnergyPlus/releases/tag/FileSystem_Move_v5

jmarrec committed:
Merge remote-tracking branch 'origin/develop' into 8376_FileSystem_Move

jmarrec committed:
Open the ssz/zsz file as a stringstream in unit tests

jmarrec committed:
adjust new unit test (added since last merge of develop into branch)

jmarrec commented:
I can reproduce the windows failing test locally if I run several tests in // that call ManageSizing.

Single test: this works

```
$ ctest -C Debug -R DesiccantDehum_VSCoolingCoilOnPrimaryAirSystemTest
```

This fails:
```
$ ctest -C Debug  -j 6 -R '(DesiccantDehum_|WindowAC_VStest1)'
```

With the added file.ssz.open_as_stringstream() in https://github.com/NREL/EnergyPlus/pull/8480/commits/0dabeb010903c36a37215ad0d1a229941c3feedb it seems to work.

jmarrec committed:
extraneous slash in the unit test I adjusted in 414b400d98bfe4165f2bdfbb2ef9a66b1c973742

jmarrec committed:
Merge remote-tracking branch 'origin/develop' into 8376_FileSystem_Move

Myoldmopar commented:
Looks so clean!!!  I'll take a final look and see if I can get this merged in.

Myoldmopar commented on .decent_ci-MacOS.yaml:
```diff
@@ -1,3 +1,3 @@
 compilers:
   - name: clang
-    cmake_extra_flags: -DCMAKE_OSX_DEPLOYMENT_TARGET=10.13 -DBUILD_FORTRAN=ON -DBUILD_TESTING:BOOL=ON -DENABLE_REGRESSION_TESTING:BOOL=ON -DREGRESSION_BASELINE_PATH:PATH=$REGRESSION_BASELINE -DREGRESSION_SCRIPT_PATH:PATH=$REGRESSION_DIR -DREGRESSION_BASELINE_SHA:STRING=$REGRESSION_BASELINE_SHA -DCOMMIT_SHA=$COMMIT_SHA -DENABLE_GTEST_DEBUG_MODE:BOOL=OFF -DLINK_WITH_PYTHON=ON
+    cmake_extra_flags: -DCMAKE_OSX_DEPLOYMENT_TARGET=10.15 -DBUILD_FORTRAN=ON -DBUILD_TESTING:BOOL=ON -DENABLE_REGRESSION_TESTING:BOOL=ON -DREGRESSION_BASELINE_PATH:PATH=$REGRESSION_BASELINE -DREGRESSION_SCRIPT_PATH:PATH=$REGRESSION_DIR -DREGRESSION_BASELINE_SHA:STRING=$REGRESSION_BASELINE_SHA -DCOMMIT_SHA=$COMMIT_SHA -DENABLE_GTEST_DEBUG_MODE:BOOL=OFF -DLINK_WITH_PYTHON=ON
```
Note to self for later release testing.

Myoldmopar commented on src/EnergyPlus/DataStringGlobals.hh:
```diff
@@ -82,68 +83,71 @@ namespace DataStringGlobals {
+    fs::path exeDirectoryPath;
+    fs::path ProgramPath;          // Path for Program from INI file
+    fs::path CurrentWorkingFolder; // Current working directory for run
+    fs::path FullPath;             // Full name of file to open, including path
```
:heart_eyes:

Myoldmopar reviewed (approved):
Built and tested locally, all good.  This is awesome and ready to go in.  Thanks for powering through @jmarrec !
